### ISSUE TITLE
`io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1OnReady` occupy too much memory

### ISSUE THREAD

**Issue by lulu2panpan** [grpc/grpc-java#11770]
my grpc verseion is :<grpc.version>1.30.0</grpc.version>
my question is ：io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1OnReady occupy too much memory

```
 num     #instances         #bytes  class name (module)
-------------------------------------------------------
   1:        166890     1899695256  [B (java.base@11.0.19.19-AJDK)
   2:      56705537     1360932888  java.util.concurrent.ConcurrentLinkedQueue$Node (java.base@11.0.19.19-AJDK)
   3:      56705251     1360926024  io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1OnReady
   4:         29061       26383648  [Ljava.lang.Object; (java.base@11.0.19.19-AJDK)
   5:         13494        8852064  io.grpc.netty.shaded.io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueue
   6:           857        4990072  [C (java.base@11.0.19.19-AJDK)
   7:         28351        2590160  [J (java.base@11.0.19.19-AJDK)
   8:         93460        2243040  java.lang.String (java.base@11.0.19.19-AJDK)
   9:         28646        1833344  io.grpc.netty.shaded.io.netty.buffer.PoolSubpage
  10:         65536        1572864  com.aliyun.polardbx.binlog.dumper.dump.logfile.parallel.EventData
  11:         12101        1445832  java.lang.Class (java.base@11.0.19.19-AJDK)
  12:         42613        1363616  java.util.concurrent.ConcurrentHashMap$Node (java.base@11.0.19.19-AJDK)
```

<img width="751" alt="截屏2024-12-21 22 17 11" src="https://github.com/user-attachments/assets/2d219b70-435d-43c9-acca-3a9dcdf587ca" />
<img width="879" alt="截屏2024-12-21 22 17 28" src="https://github.com/user-attachments/assets/76a0a665-fb8d-485d-a753-80eb89db3e37" />

**Comment by lulu2panpan** [comment]
Haha, I encountered this problem three years ago，but i forgot it，should do my logic in another thread

https://github.com/grpc/grpc-java/issues/7839 
https://github.com/grpc/grpc-java/issues/7361

**Comment by ejona86** [comment]
Yes, blocking the callback thread the way you were can accumulate onReady() callbacks. There could be some ways to fix it in gRPC, but it isn't probably worth it given how pathological your approach is (you don't even learn if the RPC died, because onClose() can't be called).

You should take a look at 1.70 though (the next release), because we merged #10318 which has a blocking API for streaming. That will probably yield _much_ better results than hard-coded polling.

**Comment by ejona86** [comment]
gRPC Java 1.30 was 4.5 years old when this issue was created, so if it was a bug we'd only be interested in investigating after an upgrade. If this is was caused by blocking the callback thread, then, yeah, that happens if you block the callback thread which prevents receiving messages.

You can use [NettyServerBuilder.flowControlWindow()](https://grpc.github.io/grpc-java/javadoc/io/grpc/netty/NettyServerBuilder.html#flowControlWindow(int)) or initialFlowControlWindow() to try to reduce how much is buffered. Like setting it to 64 KiB. initialFlowControlWindow() might slow down things briefly on new connections, but then will scale up to what is needed for performance on the connection (which is maybe the same amount as now). flowControlWindow() doesn't scale, so provides a much tighter bound on memory use.
